src/reducers/computeRenderedRows.js

```diff
@@ -207,35 +211,68 @@ function computeRenderedRowOffsets(state, rowRange) {
     return;
   }
 
-  const bufferMapping = []; // state.rows
+  // output for this function
+  const bufferMapping = state.rows;
   const rowOffsetsCache = {}; // state.rowHeights
+
+  // incremental way for calculating rowOffset
   let runningOffset = rowOffsets.sumUntil(firstBufferIdx);
-  for (let rowIdx = firstBufferIdx; rowIdx < endBufferIdx; rowIdx++) {
 
+  for (let rowIdx = firstBufferIdx; rowIdx < endBufferIdx; rowIdx++) {
     // Update the offset for rendering the row
     rowOffsetsCache[rowIdx] = runningOffset;
     runningOffset += storedHeights[rowIdx];
 
-    // Check if row already has a position in the buffer
-    let rowPosition = bufferSet.getValuePosition(rowIdx);
-
-    // Request a position in the buffer through eviction of another row
-    if (rowPosition === null && bufferSet.getSize() >= renderedRowsCount) {
-      rowPosition = bufferSet.replaceFurthestValuePosition(
-        firstViewportIdx,
-        endViewportIdx - 1,
-        rowIdx
-      );
-    }
-
-    // If we can't reuse any existing position, create a new one
-    if (rowPosition === null) {
-      rowPosition = bufferSet.getNewPositionForValue(rowIdx);
+    // we'll skip getting a row position for the rows outside the viewport
+    if (computeWithinViewportOnly && (rowIdx < firstViewportIdx || rowIdx >= endViewportIdx)) {
+      continue;
     }
 
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, bufferSet, firstViewportIdx, endViewportIdx, renderedRowsCount);
     bufferMapping[rowPosition] = rowIdx;
   }
 
+  if (computeWithinViewportOnly) {
+    // In the above loop, we only calculated row offsets for the rows inside the buffer.
+    // Now we calculate the row offsets for the remaining rows.
+    const rowsNotInBuffer = filter(state.rows, (rowIdx) => !inRange(rowIdx, firstBufferIdx, endBufferIdx));
+    rowsNotInBuffer.forEach(rowIdx => rowOffsetsCache[rowIdx] = state.rowHeights[rowIdx]);
```

Comment:
This seems incorrect to me.  Shouldn't the rowOffsetsCache be the full offest row positioning, not the rows height?  

In theory you could just have all these be offset off screen anywhere though as long as they're not in the viewport.  This seems a weird way to do that though.  

It does solve the concern I raise above about having different # of items in bufferMapping & rowOffsetsCache though...

Reply:
Ah okay let's definitely file the ticket for renaming to rowOffsets to cleanup.  

Could you walk me through how this ensures that a row from rowsNotInBuffer won't have an offset from the previous scroll which places into the buffer (or viewport)?  

The scenario I'm concerned about might look like: 
Row X is rendered at position 200, and I'm scrolled so 250 is the end of the buffer.
I scroll up so 210 is at the end of the buffer, but in scrolling up, I find the rows newly added to the beginning of the buffer are larger than expected and Row X is actually positioned at 220.
Could Row X be out of the buffer, but positioned in the viewport since it uses the old positioning of 200 instead of the new one of 220?